src/auth.ts

```diff
@@ -1,69 +1,136 @@
-import CredentialsProvider from "next-auth/providers/credentials";
+import sendgrid from "@sendgrid/mail";
+import { getServerSession, type NextAuthOptions } from "next-auth";
+import EmailProvider from "next-auth/providers/email";
 import SlackProvider from "next-auth/providers/slack";
-import { signIn as originalSignIn, type SignInOptions } from "next-auth/react";
 
-import { devMode } from "~/src/utils";
+import { authDatabaseAdapter, getUserByEmail } from "~/src/data/auth";
+import { Route } from "~/src/routing";
+
+export type OurUser = {
+  id: string;
+  slackId: string;
+  name: string;
+  email: string;
+  image: string;
+};
+
+/** NextAuth options used to configure various parts of the authentication machinery */
+export const authOptions: NextAuthOptions = {
+  adapter: authDatabaseAdapter,
+
+  providers: [
+    // OAuth flow with a Slack account, only works with the Česko.Digital workspace.
+    SlackProvider({
+      clientId: process.env.SLACK_CLIENT_ID!,
+      clientSecret: process.env.SLACK_CLIENT_SECRET!,
+      authorization: {
+        params: {
+          team: "TG21XF887",
+        },
+      },
+    }),
+    // E-mail flow, the e-mail has to already exist in our DB
+    EmailProvider({
+      server: process.env.EMAIL_SERVER,
+      from: "Česko.Digital <pomoc@cesko.digital>",
+      async sendVerificationRequest({ identifier: email, url }) {
+        sendgrid.setApiKey(process.env.SENDGRID_API_KEY ?? "");
+        const [response] = await sendgrid.send({
+          to: email,
+          from: "Česko.Digital <pomoc@cesko.digital>",
+          subject: "Přihlášení do aplikace Česko.Digital",
+          text: renderSignInEmail(url),
+          trackingSettings: {
+            clickTracking: {
+              enable: false,
+            },
+          },
+        });
+        // TBD: Do we want to check the code?
+        console.log(`Login link mailed, SendGrid response code ${response}.`);
+      },
+    }),
+  ],
+
+  callbacks: {
+    async signIn({ user }) {
+      if (user.email) {
+        const existingUser = await getUserByEmail(user.email);
+        // If user does not exist already, redirect back to sign-in page
+        // with an error explainer and the option to register instead.
+        if (!existingUser) {
+          return `/auth/sign-in?error=UserNotFound&email=${user.email}`;
+        }
+        return true; // proceed
+      } else {
+        return Route.register;
+      }
+    },
+
+    // This is called when the JWT token is created or updated. We use
+    // the callback to store the Slack ID in the token so that it is available
+    // in the session data on the client. TBD: This is a legacy hack, we
+    // should replace the Slack ID with the regular database ID of the user.
+    async jwt({ token, user }) {
+      if (user) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        token.slackId = (user as any as OurUser).slackId;
```

Comment:
V TypeScriptu muzes pre-typovat modul takto (v samostatnem souboru):

```
declare module "next-auth" {
  interface Session {
    user: OurUser;
  }
}
```

Reply:
Díky! Oni k tomu mají dokonce [celou sekci v nápovědě](https://next-auth.js.org/getting-started/typescript#module-augmentation), zkoušel jsem to, ale mělo to hromadu ostrých hran, tak jsem se na to nakonec vykašlal výměnou za tahle prasečí přetypování.